lib/datasource/crate/index.ts

```diff
@@ -1,43 +1,168 @@
+  clonePath?: string;
+}
+
+export function getIndexSuffix(lookupName: string): string[] {
   const len = lookupName.length;
 
   if (len === 1) {
-    return '1/' + lookupName;
+    return ['1', lookupName];
   }
   if (len === 2) {
-    return '2/' + lookupName;
+    return ['2', lookupName];
   }
   if (len === 3) {
-    return '3/' + lookupName[0] + '/' + lookupName;
+    return ['3', lookupName[0], lookupName];
   }
 
-  return (
-    lookupName.slice(0, 2) + '/' + lookupName.slice(2, 4) + '/' + lookupName
-  );
+  return [lookupName.slice(0, 2), lookupName.slice(2, 4), lookupName];
 }
 
 interface CrateRecord {
   vers: string;
   yanked: boolean;
 }
 
+async function fetchCrateRecordsPayload(
+  info: RegistryInfo,
+  lookupName: string
+): Promise<string> {
+  if (info.clonePath) {
+    const path = join(info.clonePath, ...getIndexSuffix(lookupName));
+    return fs.readFile(path, { encoding: 'utf8' });
+  }
+
+  if (info.flavor === RegistryFlavor.CratesIo) {
+    const crateUrl = CRATES_IO_BASE_URL + getIndexSuffix(lookupName).join('/');
+    try {
+      return (await http.get(crateUrl)).body;
+    } catch (err) {
+      if (
+        err.statusCode === 429 ||
+        (err.statusCode >= 500 && err.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
+      }
+    }
+  }
+
+  throw new Error(`unsupported crate registry flavor: ${info.flavor}`);
+}
+
+/// Computes the dependency URL for a crate, given
+/// registry information
+function getDependencyUrl(info: RegistryInfo, lookupName: string): string {
+  switch (info.flavor) {
+    case RegistryFlavor.CratesIo:
+      return `https://crates.io/crates/${lookupName}`;
+    case RegistryFlavor.Cloudsmith: {
+      // input: https://dl.cloudsmith.io/basic/$org/$repo/cargo/index.git
+      const tokens = info.url.pathname.split('/');
+      const org = tokens[2];
+      const repo = tokens[3];
+      return `https://cloudsmith.io/~${org}/repos/${repo}/packages/detail/cargo/${lookupName}`;
+    }
+    default:
+      return `${info.rawUrl}/${lookupName}`;
+  }
+}
+
+/// Fetches information about a registry, by url.
+/// If no url is given, assumes crates.io.
+/// If an url is given, assumes it's a valid Git repository
+/// url and clones it to cache.
+async function fetchRegistryInfo(
+  registryUrl: string
+): Promise<RegistryInfo | null> {
+  let url: URL;
+  try {
+    url = new URL(registryUrl);
+  } catch (err) {
+    logger.debug({ registryUrl }, 'could not parse registry URL');
+    return null;
+  }
+
+  let flavor: RegistryFlavor;
+  if (url.hostname === 'crates.io') {
+    flavor = RegistryFlavor.CratesIo;
+  } else if (url.hostname === 'dl.cloudsmith.io') {
+    flavor = RegistryFlavor.Cloudsmith;
+  } else {
+    flavor = RegistryFlavor.Other;
+  }
+
+  const registry: RegistryInfo = {
+    flavor,
+    rawUrl: registryUrl,
+    url,
+  };
+
+  if (flavor !== RegistryFlavor.CratesIo) {
+    let clonePath = registryClonePaths[registryUrl];
```

Comment:
This avoids re-cloning registries over and over again on a single renovate run.

I have no doubt there's a better strategy available / one that fits existing renovate patterns better, I just haven't found one yet - suggestions welcome.

Reply:
Note that the built-in renovate file+mem cache is not used here, but maybe it should be? I was worried it would miss updates, but I see there's a 10-minute TTL on some other cache entries.